targets/TARGET_STM/cpu_uid_api.c

```diff
@@ -0,0 +1,151 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *******************************************************************************
+ */
+#include "device.h"
+
+#if DEVICE_CPUUID
+
+#include <string.h>
+#if defined(TARGET_STM32F0)
+#include "stm32f0xx_ll_utils.h"
+#elif defined(TARGET_STM32F1)
+#include "stm32f1xx_ll_utils.h"
+#elif defined(TARGET_STM32F2)
+#include "stm32f2xx_ll_utils.h"
+#elif defined(TARGET_STM32F3)
+#include "stm32f3xx_ll_utils.h"
+#elif defined(TARGET_STM32F4)
+#include "stm32f4xx_ll_utils.h"
+#elif defined(TARGET_STM32F7)
+#include "stm32f7xx_ll_utils.h"
+#elif defined(TARGET_STM32L0)
+#include "stm32l0xx_ll_utils.h"
+#elif defined(TARGET_STM32L1)
+#include "stm32l1xx_ll_utils.h"
+#elif defined(TARGET_STM32L4)
+#include "stm32l4xx_ll_utils.h"
+#else
+#error "Unsupported STM32 device!"
+#endif
+#include "cpu_uid_api.h"
+
+#define UID_WORDS       3
+
+#ifdef MBED_CPU_UID_STR_SIZE_MAX
+static int cpu_uid_itoa(int value, char *sp, int radix);
+#endif
+
+void cpu_uid_get_uid(uint8_t *uid)
+{
+    uint32_t uid_buf[UID_WORDS];
+    int pos = 0;
+
+    uid_buf[0] = LL_GetUID_Word0();
+    uid_buf[1] = LL_GetUID_Word1();
+    uid_buf[2] = LL_GetUID_Word2();
+    
+    for (int i = (UID_WORDS-1); i >= 0; --i) {
+        for (int j = 3; j >= 0; --j) {
+            uid[pos++] = (uint8_t)((uid_buf[i] >> (j*8)) & 0xFF);
+        }
+    }
+}
+
+#ifdef MBED_CPU_UID_STR_SIZE_MAX
+void cpu_uid_get_str(char *str)
+{
+    char buffer[16];
+    char lot_number[8];
+    uint8_t wafer_number;
+    uint16_t x, y;
+    uint32_t uid_buf[UID_WORDS];
+
+    uid_buf[0] = LL_GetUID_Word0();
+    uid_buf[1] = LL_GetUID_Word1();
+    uid_buf[2] = LL_GetUID_Word2();
+
+    lot_number[0] = (uid_buf[2] >> 24) & 0x000000FF;
+    lot_number[1] = (uid_buf[2] >> 16) & 0x000000FF;
+    lot_number[2] = (uid_buf[2] >> 8) & 0x000000FF;
+    lot_number[3] = uid_buf[2] & 0x000000FF;
+    lot_number[4] = (uid_buf[1] >> 24) & 0x000000FF;
+    lot_number[5] = (uid_buf[1] >> 16) & 0x000000FF;
+    lot_number[6] = (uid_buf[1] >> 8) & 0x000000FF;
+    lot_number[7] = '\0';
+    
+    wafer_number = uid_buf[1] & 0x000000FF;
+    
+    x = (uid_buf[0] >> 16) & 0x0000FFFF;
+    y = uid_buf[0] & 0x0000FFFF;
+    
+    // make a string in format "Lot xxxxxxx Wafer xxx X=xxxxx Y=xxxxx" - max 38 chars including terminating zero
```

Comment:
This is absolutely fine, but I'd been thinking of something a little more compact:
`LLLLLLL:WWW:XXXXX:YYYYY`